pygmt/src/grd2xyz.py

```diff
@@ -0,0 +1,86 @@
+"""
+grd2xyz - Convert grid to data table
+"""
+import pandas as pd
+from pygmt.clib import Session
+from pygmt.exceptions import GMTInvalidInput
+from pygmt.helpers import (
+    GMTTempFile,
+    build_arg_string,
+    fmt_docstring,
+    kwargs_to_strings,
+    use_alias,
+)
+
+
+@fmt_docstring
+@use_alias(
+    R="region",
+    V="verbose",
+)
+@kwargs_to_strings(R="sequence")
+def grd2xyz(grid, output_type="pandas", outfile=None, **kwargs):
+    r"""
+    Convert grid to data table.
+
+    Read a grid and output xyz-triplets as a :class:`numpy.ndarray`,
+    :class:`pandas.DataFrame`, or ASCII file.
+
+    Full option list at :gmt-docs:`grd2xyz.html`
+
+    {aliases}
+
+    Parameters
+    ----------
+    grid : str or xarray.DataArray
+        The file name of the input grid or the grid loaded as a
+        :class:`xarray.DataArray`. This is the only required parameter.
+    output_type : str
+        Determine the format the xyz data will be returned in [Default is
+        ``pandas``]:
+
+            - ``numpy`` - :class:`numpy.ndarray`
+            - ``pandas``-  :class:`pandas.DataFrame`
+            - ``file`` - ASCII file (requires ``outfile``)
+    outfile : str
+        The file name for the output ASCII file.
+    {R}
+    {V}
+
+    Returns
+    -------
+    ret : pandas.DataFrame or numpy.ndarray or None
+        Return type depends on ``outfile`` and ``output_type``:
+
+        - None if ``outfile`` is set (output will be stored in file set by
+          ``outfile``)
+        - :class:`pandas.DataFrame` or :class:`numpy.ndarray` if ``outfile`` is
+          not set (depends on ``output_type`` [Default is
+          :class:`pandas.DataFrame`])
+
+    """
+    if output_type not in ["numpy", "pandas", "file"]:
+        raise GMTInvalidInput(
+            """Must specify format as either numpy, pandas, or file."""
+        )
+    if output_type == "file" and outfile is None:
+        raise GMTInvalidInput("""Must specify outfile for ASCII output.""")
```

Comment:
> > @meghanrjones I think my preference is for 3, with the only addition being that a warning will be raised that states that the `output_type` is being changed to "file" because an argument was passed to `output_file` so that it is made explicitly clear to the user. My rationale is that it is simpler to have different parameters accomplish different things (one to set the output type, one to name it) rather than having a single parameter that makes multiple changes (sets output type, creates a filename).
> 
> I agree with @willschlitzer.

If we're going for option 3 and adding the warning, then try something like this (debating whether to use `SyntaxWarning` or `RuntimeWarning`, xref https://docs.python.org/3/library/warnings.html#warning-categories and #1316).

```suggestion
    if outfile is not None and output_type != "file":
        msg = (
            f"Changing `output_type` of grd2xyz from '{output_type}' to 'file' "
            "since `outfile` parameter is set. Please use `output_type='file'` "
            "to silence this warning."
        )
        warnings.warn(msg, category=RuntimeWarning, stacklevel=2)
    if output_type == "file" and outfile is None:
        raise GMTInvalidInput("""Must specify outfile for ASCII output.""")
```

Reply:
@weiji14 Why would this be different than just using a `GMTInvalidInput` error? I agree that this falls under the description of a `RuntimeWarning` as described in the Python docs, but it's not clear to me why it isn't better suited for `GMTInvalid Input`.

Reply:
The `GMTInvalidInput` is blocking, whereas the `RuntimeWarning` is non-blocking (i.e. the module will continue to run). Did you want to make this an error or just a warning?

Reply:
Makes sense. It looks like it will still output to a file as long as an `outfile` name is set, regardless of the argument passed to `output_type`.

Reply:
@weiji14 I'm struggling to write a test for this warning. This is what I have so far:

```def test_grd2xyz_outfile_incorrect_output_type(grid):
    """
    Test that grd2xyz raises a warning when an outfile name is set but the
    output_type is not set to file is set with no outfile.
    """
    with pytest.raises(RuntimeWarning):
        with GMTTempFile(suffix=".xyz") as tmpfile:
            result = grd2xyz(grid=grid, outfile=tmpfile.name, output_type="numpy")
```

It returns `Failed: DID NOT RAISE <class 'RuntimeWarning'>`. 

I'm looking through the [pytest warning documentation](https://docs.pytest.org/en/6.2.x/warnings.html) but not finding something that describes this type of test.